src/ripple/basics/base_uint.h

```diff
@@ -126,25 +126,45 @@ class base_uint
         *this = beast::zero;
     }
 
-    explicit base_uint (Blob const& vch)
+    explicit base_uint (std::uint64_t b)
     {
-        assert (vch.size () == size ());
-
-        if (vch.size () == size ())
-            memcpy (pn.data (), vch.data (), size ());
-        else
-            *this = beast::zero;
+        *this = b;
     }
 
-    explicit base_uint (std::uint64_t b)
+    template <class Container,
+              class = std::void_t
+              <
+                  decltype(std::declval<Container const>().size()),
```

Comment:
Are these like a poor man's concepts? I'm not opposed to them, but just to confirm my understanding of C++, they aren't strictly necessary for correct compilation, are they?

Reply:
Yes, poor man's concepts.  They are necessary to prevent _incorrect_ compilation, such as accidentally constructing from a range of types that shouldn't be `memcpy`'d.

Reply:
Could these be pulled into their own type_trait? :) 
```
template<typename Container>
struct void_contiguous_container_t : std::void_t<
    decltype(std::declval<Container const>().size()),
    decltype(std::declval<Container const>().data()),
    typename Container::value_type
>{};

template<typename Container, typename Type = void>
using enable_if_trivially_copyable_t = std::enable_if_t<
    std::is_trivially_copyable<typename Container::value_type>::value, Type
>;
```

Then you can use it like... 

```
    template <class Container,
              class = void_contiguous_container_t<Container>,
              class = enable_if_trivially_copyable_t<Container>>
    explicit base_uint(Container const& c)
    {
        assert(c.size()*sizeof(typename Container::value_type) == size());
        std::memcpy(data_.data(), c.data(), size());
    }
```

I like giving names to custom traits, layering on the core type_traits. It's verbose, but doesn't require comments to explain it/less copy paste. The names I picked probably aren't the best, but that's the basic idea.

Reply:
We could.  I'd be more into it if we needed `void_contiguous_container_t` more places and more often.  With it just in `base_uint` I don't see a benefit to create this and move its definition off to a generic utility file.  However if people strongly feel that would make the code better, I don't mind doing it.

Reply:
My limit is usually ~1 duplication, but I know others are less bothered by it. It also doesn't need to be moved to a different file either, I usually just stick little utilities like these in an anonymous namespace at the top of the file after `#includes`. Then if it's used in multiple places, pull it out into a header.

It's mostly about readability and understanding the intent. I can read `contiguous_container_t` (or maybe `contiguous_container_concept` ...?) and immediately understand what the intent is, vs. reading those 3 lines and having to guess what the intent is. But it's just an idea 😃, I wouldn't block on it.

Reply:
@HowardHinnant I was under the impression that the `std::is_trivially_copable` constraint is necessary for correct compilation, but that the member checks are not. I thought that template instantiation would just fail if those members were absent. Is that correct? I guess I should have been more clear that my comment was for the `std::void_t` expression only, not the `std::is_trivially_copyable` expression.

I weakly prefer @jwbusch's suggestion for a named trait, too, but I'm not going to block on it either.

Reply:
The constraints keep other constructors such as `base_uint (std::uint64_t b)` accidentally getting confused with the contiguous container constructor (and ditto for assignment).  I did not test that the constraints were necessary.  Just I've hit that bug so many times I've developed an aversion to "overly generic" signatures.

I'm testing a contiguous_container trait and will push it once I'm confident in it.

Reply:
And pushed.

Reply:
I think the non-template always wins as more-specific (see [tie-breaker #4 under "Best viable function"](https://en.cppreference.com/w/cpp/language/overload_resolution)), and my [limited test](https://wandbox.org/permlink/mDXQk8DnKmmTOirY) suggests as much, but I'm totally fine with erring on the side of caution.